File: examples/form-comparison/servers/rust/src/repository.rs
use crate::{
    bad,
    form::{collection, rows},
    json as codec, Result,
};
use serde_json::{json, Map, Value};
use std::{
    collections::HashSet,
    fs::{self, OpenOptions},
    io::Write,
    path::{Path, PathBuf},
};

pub struct Repository {
    pub file: PathBuf,
    pub fixtures: PathBuf,
}

pub fn read_object(path: &Path) -> Result<Value> {
    let value = codec::decode(&fs::read(path)?)?;
    if !value.is_object() {
        return Err(bad("Expected stored object"));
    }
    Ok(value)
}

impl Repository {
    fn fixture(&self, name: &str) -> Result<Value> {
        read_object(&self.fixtures)?
            .get(name)
            .cloned()
            .ok_or_else(|| bad("Unknown fixture"))
    }

    /// Lock the file before reading and atomically replace changed records.
    fn transaction(
        &self,
        operation: impl FnOnce(&Value) -> Result<(Value, Value)>,
    ) -> Result<Value> {
        let lock = OpenOptions::new()
            .create(true)
            .truncate(false)
            .read(true)
            .write(true)
            .open(self.file.with_extension("json.lock"))?;
        lock.lock()?;
        let before = match fs::read(&self.file) {
            Ok(bytes) => codec::decode(&bytes)?,
            Err(error) if error.kind() == std::io::ErrorKind::NotFound => {
                self.fixture("default")?
            }
            Err(error) => return Err(error.into()),
        };
        let (after, result) = operation(&before)?;
        let encoded = codec::encode(&after)?;
        if !self.file.exists() || codec::encode(&before)? != encoded {
            let mut temporary = tempfile::NamedTempFile::new_in(
                self.file
                    .parent()
                    .ok_or_else(|| bad("Missing repository directory"))?,
            )?;
            writeln!(temporary, "{encoded}")?;
            temporary.as_file().sync_all()?;
            temporary.persist(&self.file).map_err(|e| e.error)?;
        }
        Ok(result)
    }

    pub fn read(&self) -> Result<Value> {
        self.transaction(|value| Ok((value.clone(), value.clone())))
    }
    pub fn reset(&self, name: &str) -> Result<Value> {
        let state = self.fixture(name)?;
        self.transaction(|_| Ok((state.clone(), state)))
    }

    /// Assign sequences, enforce ownership and return scoped key changes.
    pub fn save(&self, data: &Value) -> Result<Value> {
        self.transaction(|before| {
            let mut next = before["next"].clone();
            let (mut companies,mut stores,mut departments,mut changes) = (Vec::new(),Vec::new(),Vec::new(),Vec::new());
            let (mut seen_company,mut seen_store,mut seen_department) = (HashSet::new(),HashSet::new(),HashSet::new());
            for (company_key,company) in rows(&data["companies"])? {
                let company_seq = allocate(&company_key,&table(before,"companies")?.iter().collect::<Vec<_>>(),"company",&mut next,&mut seen_company)?;
                companies.push(json!({"company_seq":company_seq,"name":company["name"],"position":companies.len()}));
                for (position,(store_key,store)) in rows(&company["stores"])?.into_iter().enumerate() {
                    check_parent(&store_key,table(before,"stores")?,"store","company",&company_seq)?;
                    let store_seq = allocate(&store_key,&owned(table(before,"stores")?,"company",&company_seq),"store",&mut next,&mut seen_store)?;
                    stores.push(json!({"store_seq":store_seq,"company_seq":company_seq,"position":position,"name":store["name"],"enabled":store["enabled"],"detail":store["detail"],"title":store["title"]}));
                    for (position,(department_key,department)) in rows(&store["departments"])?.into_iter().enumerate() {
                        check_parent(&department_key,table(before,"departments")?,"department","store",&store_seq)?;
                        let department_seq = allocate(&department_key,&owned(table(before,"departments")?,"store",&store_seq),"department",&mut next,&mut seen_department)?;
                        departments.push(json!({"department_seq":department_seq,"store_seq":store_seq,"position":position,"name":department["name"]}));
                        change(&mut changes,&format!("companies.{company_key}.stores.{store_key}.departments"),&department_key,&department_seq)?;
                    }
                    change(&mut changes,&format!("companies.{company_key}.stores"),&store_key,&store_seq)?;
                }
                change(&mut changes,"companies",&company_key,&company_seq)?;
            }
            let after = json!({"next":next,"companies":companies,"stores":stores,"departments":departments});
            let loaded = load_data(&after)?;
            Ok((after.clone(),json!({"storage":after,"data":loaded,"keyChanges":changes})))
        })
    }
}

fn table<'a>(state: &'a Value, kind: &str) -> Result<&'a Vec<Value>> {
    state[kind]
        .as_array()
        .ok_or_else(|| bad("Expected stored rows"))
}
fn sequence<'a>(row: &'a Value, kind: &str) -> Result<&'a str> {
    row[format!("{kind}_seq")]
        .as_str()
        .ok_or_else(|| bad("Expected stored sequence"))
}

fn key(sequence: &str) -> Result<String> {
    if sequence.is_empty() || sequence.len() > 13 || !sequence.bytes().all(|b| b.is_ascii_digit()) {
        return Err(bad("Invalid stored sequence"));
    }
    Ok(format!(
        "__{}{}__",
        "0".repeat(13 - sequence.len()),
        sequence
    ))
}

fn ordered_rows(rows: Vec<&Value>) -> Result<Vec<&Value>> {
    let mut entries: Vec<_> = rows
        .into_iter()
        .map(|row| {
            Ok((
                row["position"]
                    .as_u64()
                    .ok_or_else(|| bad("Expected stored position"))?,
                row,
            ))
        })
        .collect::<Result<_>>()?;
    entries.sort_by_key(|(position, _)| *position);
    Ok(entries.into_iter().map(|(_, row)| row).collect())
}

fn owned<'a>(rows: &'a [Value], parent: &str, seq: &str) -> Vec<&'a Value> {
    rows.iter()
        .filter(|row| row[format!("{parent}_seq")].as_str() == Some(seq))
        .collect()
}

fn row_object(value: Value) -> Map<String, Value> {
    value
        .as_object()
        .expect("row constructor creates an object")
        .clone()
}

/// Reconstruct row order from positions rather than physical table order.
pub fn load_data(state: &Value) -> Result<Value> {
    let mut companies = Vec::new();
    for company in ordered_rows(table(state, "companies")?.iter().collect())? {
        let company_seq = sequence(company, "company")?;
        let mut stores = Vec::new();
        for store in ordered_rows(owned(table(state, "stores")?, "company", company_seq))? {
            let store_seq = sequence(store, "store")?;
            let mut departments = Vec::new();
            for department in ordered_rows(owned(table(state, "departments")?, "store", store_seq))?
            {
                let seq = sequence(department, "department")?;
                departments.push((key(seq)?, row_object(json!({"name":department["name"]}))));
            }
            let value = json!({"name":store["name"],"enabled":store["enabled"],"detail":store["detail"],"title":store["title"],"departments":collection(departments)});
            stores.push((key(store_seq)?, row_object(value)));
        }
        let value = json!({"name":company["name"],"stores":collection(stores)});
        companies.push((key(company_seq)?, row_object(value)));
    }
    Ok(json!({"companies":collection(companies)}))
}

fn allocate(
    row_key: &str,
    existing: &[&Value],
    kind: &str,
    next: &mut Value,
    seen: &mut HashSet<String>,
) -> Result<String> {
    let field = format!("{kind}_seq");
    let mut seq = None;
    for candidate in existing {
        let saved = sequence(candidate, kind)?;
        if key(saved)? == row_key {
            seq = Some(saved.to_string());
            break;
        }
    }
    let seq = if let Some(seq) = seq {
        if seen.contains(&seq) {
            return Err(bad(format!("Duplicate {field}: {seq}")));
        }
        seq
    } else {
        let number = next[kind]
            .as_u64()
            .ok_or_else(|| bad("Expected stored sequence counter"))?;
        let seq = number.to_string();
        key(&seq)?;
        next[kind] = json!(number + 1);
        seq
    };
    seen.insert(seq.clone());
    Ok(seq)
}

fn check_parent(
    row_key: &str,
    rows: &[Value],
    kind: &str,
    parent: &str,
    parent_seq: &str,
) -> Result<()> {
    for row in rows {
        if key(sequence(row, kind)?)? == row_key && sequence(row, parent)? != parent_seq {
            return Err(bad(format!("Incorrect parent for {kind}_seq")));
        }
    }
    Ok(())
}

fn change(changes: &mut Vec<Value>, path: &str, old: &str, sequence: &str) -> Result<()> {
    let new = key(sequence)?;
    if old != new {
        changes.push(json!({"path":path,"oldKey":old,"newKey":new}));
    }
    Ok(())
}

File: examples/form-comparison/servers/go/repository.go
package main

import (
	"bytes"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strconv"
	"strings"
	"syscall"
)

type repository struct{ file, fixtures string }
type inputError string

func (e inputError) Error() string { return string(e) }

func readObject(file string) (*object, error) {
	bytes, err := os.ReadFile(file)
	if err != nil {
		return nil, err
	}
	value, err := decodeJSON(bytes)
	if err != nil {
		return nil, err
	}
	result, ok := value.(*object)
	if !ok {
		return nil, fmt.Errorf("Expected stored object")
	}
	return result, nil
}

func (r repository) fixture(name string) (*object, error) {
	fixtures, err := readObject(r.fixtures)
	if err != nil {
		return nil, err
	}
	value, ok := get(fixtures, name).(*object)
	if !ok {
		return nil, inputError("Unknown fixture")
	}
	return value, nil
}

// lockedFile runs operation on the decoded contents of file under an exclusive lock, seeding a
// missing file, and atomically replaces the file when the missing file was seeded or the contents
// changed. An unreadable or malformed file fails and is left as it is.
func lockedFile(file string, seed func() (any, error), operation func(any) (any, any, error)) (any, error) {
	lock, err := os.OpenFile(file+".lock", os.O_CREATE|os.O_RDWR, 0600)
	if err != nil {
		return nil, err
	}
	defer lock.Close()
	if err = syscall.Flock(int(lock.Fd()), syscall.LOCK_EX); err != nil {
		return nil, err
	}
	defer syscall.Flock(int(lock.Fd()), syscall.LOCK_UN)
	var before any
	stored, err := os.ReadFile(file)
	missing := os.IsNotExist(err)
	switch {
	case missing:
		before, err = seed()
	case err == nil:
		before, err = decodeJSON(stored)
	}
	if err != nil {
		return nil, err
	}
	after, result, err := operation(before)
	if err != nil {
		return nil, err
	}
	encoded, err := encodeJSON(after)
	if err != nil {
		return nil, err
	}
	encoded = append(encoded, '\n')
	if !missing && bytes.Equal(stored, encoded) {
		return result, nil
	}
	temp, err := os.CreateTemp(filepath.Dir(file), ".form-")
	if err != nil {
		return nil, err
	}
	name := temp.Name()
	defer os.Remove(name)
	if _, err = temp.Write(encoded); err != nil {
		temp.Close()
		return nil, err
	}
	if err = temp.Sync(); err != nil {
		temp.Close()
		return nil, err
	}
	if err = temp.Close(); err != nil {
		return nil, err
	}
	if err = os.Rename(name, file); err != nil {
		return nil, err
	}
	return result, nil
}

// transaction locks the form file and atomically replaces changed records.
func (r repository) transaction(operation func(*object) (*object, *object, error)) (*object, error) {
	result, err := lockedFile(r.file, func() (any, error) { return r.fixture("default") },
		func(value any) (any, any, error) {
			state, ok := value.(*object)
			if !ok {
				return nil, nil, fmt.Errorf("Expected stored object")
			}
			after, result, err := operation(state)
			return after, result, err
		})
	if err != nil {
		return nil, err
	}
	return result.(*object), nil
}

func (r repository) read() (*object, error) {
	return r.transaction(func(state *object) (*object, *object, error) { return state, state, nil })
}
func (r repository) reset(name string) (*object, error) {
	state, err := r.fixture(name)
	if err != nil {
		return nil, err
	}
	return r.transaction(func(_ *object) (*object, *object, error) { return state, state, nil })
}

func sequenceKey(sequence string) (string, error) {
	if len(sequence) < 1 || len(sequence) > 13 {
		return "", fmt.Errorf("Invalid stored sequence")
	}
	for _, char := range sequence {
		if char < '0' || char > '9' {
			return "", fmt.Errorf("Invalid stored sequence")
		}
	}
	return "__" + strings.Repeat("0", 13-len(sequence)) + sequence + "__", nil
}

func integer(value any) int {
	switch value := value.(type) {
	case int:
		return value
	case float64:
		return int(value)
	}
	panic("Expected stored integer")
}
func storedRows(state *object, kind string) []any { return get(state, kind).([]any) }

func sortedRows(items []any, parent, sequence string) []*object {
	result := []*object{}
	for _, item := range items {
		row := item.(*object)
		if parent == "" || get(row, parent+"_seq") == sequence {
			result = append(result, row)
		}
	}
	sort.SliceStable(result, func(i, j int) bool { return integer(get(result[i], "position")) < integer(get(result[j], "position")) })
	return result
}

// loadData reconstructs collection order from stored positions and parent IDs.
func loadData(state *object) (*object, error) {
	companies := []row{}
	for _, company := range sortedRows(storedRows(state, "companies"), "", "") {
		companySeq := get(company, "company_seq").(string)
		stores := []row{}
		for _, store := range sortedRows(storedRows(state, "stores"), "company", companySeq) {
			storeSeq := get(store, "store_seq").(string)
			departments := []row{}
			for _, department := range sortedRows(storedRows(state, "departments"), "store", storeSeq) {
				seq := get(department, "department_seq").(string)
				key, err := sequenceKey(seq)
				if err != nil {
					return nil, err
				}
				departments = append(departments, row{key, record("name", get(department, "name"))})
			}
			key, err := sequenceKey(storeSeq)
			if err != nil {
				return nil, err
			}
			value := record("name", get(store, "name"), "enabled", get(store, "enabled"), "detail", get(store, "detail"), "title", clone(get(store, "title")), "departments", rowCollection(departments))
			stores = append(stores, row{key, value})
		}
		key, err := sequenceKey(companySeq)
		if err != nil {
			return nil, err
		}
		value := record("name", get(company, "name"), "stores", rowCollection(stores))
		companies = append(companies, row{key, value})
	}
	return record("companies", rowCollection(companies)), nil
}

func allocate(item row, existing []any, kind string, next *object, seen map[string]bool) (string, error) {
	field := kind + "_seq"
	sequence := ""
	for _, raw := range existing {
		candidate := raw.(*object)
		seq := get(candidate, field).(string)
		key, err := sequenceKey(seq)
		if err != nil {
			return "", err
		}
		if key == item.key {
			sequence = seq
			break
		}
	}
	if seen[sequence] {
		return "", inputError(fmt.Sprintf("Duplicate %s: %s", field, sequence))
	}
	if sequence == "" {
		number := integer(get(next, kind))
		sequence = strconv.Itoa(number)
		if _, err := sequenceKey(sequence); err != nil {
			return "", err
		}
		next.Set(kind, number+1)
	}
	seen[sequence] = true
	return sequence, nil
}

func owned(items []any, parent, sequence string) []any {
	result := []any{}
	for _, raw := range items {
		if get(raw.(*object), parent+"_seq") == sequence {
			result = append(result, raw)
		}
	}
	return result
}

func checkParent(key string, items []any, kind, parent, sequence string) error {
	for _, raw := range items {
		item := raw.(*object)
		saved, err := sequenceKey(get(item, kind+"_seq").(string))
		if err != nil {
			return err
		}
		if key == saved && get(item, parent+"_seq") != sequence {
			return inputError(fmt.Sprintf("Incorrect parent for %s_seq", kind))
		}
	}
	return nil
}

// save assigns sequences, preserves existing ownership and returns scoped key changes.
func (r repository) save(data *object) (*object, error) {
	return r.transaction(func(before *object) (*object, *object, error) {
		next := clone(get(before, "next")).(*object)
		companies, stores, departments, changes := []any{}, []any{}, []any{}, []any{}
		seenCompany, seenStore, seenDepartment := map[string]bool{}, map[string]bool{}, map[string]bool{}
		change := func(path, key, sequence string) error {
			saved, err := sequenceKey(sequence)
			if err != nil {
				return err
			}
			if saved != key {
				changes = append(changes, record("path", path, "oldKey", key, "newKey", saved))
			}
			return nil
		}
		companyRows, err := rows(get(data, "companies"), "companies")
		if err != nil {
			return nil, nil, err
		}
		for _, company := range companyRows {
			companySeq, err := allocate(company, storedRows(before, "companies"), "company", next, seenCompany)
			if err != nil {
				return nil, nil, err
			}
			companies = append(companies, record("company_seq", companySeq, "name", get(company.value, "name"), "position", len(companies)))
			storeRows, err := rows(get(company.value, "stores"), "stores")
			if err != nil {
				return nil, nil, err
			}
			for i, store := range storeRows {
				if err = checkParent(store.key, storedRows(before, "stores"), "store", "company", companySeq); err != nil {
					return nil, nil, err
				}
				storeSeq, err := allocate(store, owned(storedRows(before, "stores"), "company", companySeq), "store", next, seenStore)
				if err != nil {
					return nil, nil, err
				}
				stores = append(stores, record("store_seq", storeSeq, "company_seq", companySeq, "position", i, "name", get(store.value, "name"), "enabled", get(store.value, "enabled"), "detail", get(store.value, "detail"), "title", get(store.value, "title")))
				departmentRows, err := rows(get(store.value, "departments"), "departments")
				if err != nil {
					return nil, nil, err
				}
				for j, department := range departmentRows {
					if err = checkParent(department.key, storedRows(before, "departments"), "department", "store", storeSeq); err != nil {
						return nil, nil, err
					}
					departmentSeq, err := allocate(department, owned(storedRows(before, "departments"), "store", storeSeq), "department", next, seenDepartment)
					if err != nil {
						return nil, nil, err
					}
					departments = append(departments, record("department_seq", departmentSeq, "store_seq", storeSeq, "position", j, "name", get(department.value, "name")))
					if err = change("companies."+company.key+".stores."+store.key+".departments", department.key, departmentSeq); err != nil {
						return nil, nil, err
					}
				}
				if err = change("companies."+company.key+".stores", store.key, storeSeq); err != nil {
					return nil, nil, err
				}
			}
			if err = change("companies", company.key, companySeq); err != nil {
				return nil, nil, err
			}
		}
		after := record("next", next, "companies", companies, "stores", stores, "departments", departments)
		loaded, err := loadData(after)
		if err != nil {
			return nil, nil, err
		}
		return after, record("storage", after, "data", loaded, "keyChanges", changes), nil
	})
}
